keps/sig-api-machinery/4222-cbor-serializer/README.md

```diff
@@ -0,0 +1,1154 @@
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+Under this proposal, Kubernetes API servers and clients will support the Concise
+Binary Object Representation (CBOR) data format. CBOR will be available to
+clients as an alternative to JSON for serializing resources in request and
+response bodies. It will supersede JSON in apiextensions-apiserver for storage
+serialization of custom resources.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+In the course of processing a single request to the Kubernetes API, various
+representations of a resource may be encoded and decoded several times by both
+the client (encode request body, decode response body) and the server (decode
+request body, decode from storage, encode response body, encode to storage). For
+years, Kubernetes has supported a Protobuf format requiring dramatically less
+CPU time and heap churn than its JSON (or YAML) format. The reduction in codec
+overhead resulting from the adoption of Protobuf has made Kubernetes clusters
+more efficient and able to handle increasingly heavy API traffic.
+
+The Kubernetes community has embraced CustomResourceDefinitions (CRDs) as a
+declarative extension mechanism for the Kubernetes API. Unlike native types,
+custom resources can not trivially be serialized as Protobuf for serving or
+storage. Protobuf is dependent on code generation, requires careful schema
+evolution, and requires clients and servers to have compilation-time knowledge
+of any Protobuf definitions they will use.
+
+High-object-count and high-traffic custom resources are at a serious efficiency
+disadvantage versus comparable native resources. Benchmarks suggest that custom
+resource and dynamic client encode and decode operations can be made up to
+approximately 8x and 2x faster, respectively, with a substantial reduction in
+heap allocations, by adopting CBOR as a self-describing binary format.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Reduce CPU time and heap churn of encode and decode for cases that can't use
```

Comment:
Sorry, I forgot to send my comments and now I see they are out-dated, reviewing the changes.

Reply:
No problem, glad you noticed and thanks for re-reviewing!

> Are we going to change kubectl to convert from YAML to CBOR when doing apply?

Yes, but only after all versions within kubectl's supported skew accept CBOR. It will still need to fall back if patching a resource handled by an older aggregated API server.

> yaml.Marshal throughout the code base

This goal statement is broader than intended. I have been focusing for this proposal on encoding/decoding objects for request/response bodies and storage. I don't see why CBOR could not eventually be used in certain cases considered individually, where it would be interchangeable and there is enough benefit.